crates/order-validation/src/banned.rs

```diff
@@ -1,34 +1,183 @@
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+    limit_order_participant: bool,
+}
+
+impl UserMetadata {
+    pub fn with_banned(mut self, banned: bool) -> Self {
+        self.is_banned = banned;
+        self
+    }
+
+    pub fn with_last_updated(mut self, last_updated: Instant) -> Self {
+        self.last_updated = last_updated;
+        self
+    }
 }
 
 struct Onchain {
     contract: ChainalysisOracle,
-    cache: Mutex<TimedCache<H160, bool>>,
+    db_pool: PgPool,
+    cache: RwLock<HashMap<H160, UserMetadata>>,
 }
 
-const TTL: u64 = 60 * 60; // 1 hour
+impl Onchain {
+    pub fn new(contract: ChainalysisOracle, db_pool: PgPool) -> Arc<Self> {
+        let onchain = Arc::new(Self {
+            contract,
+            db_pool,
+            cache: Default::default(),
+        });
+
+        onchain.clone().spawn_maintenance_task();
+
+        onchain
+    }
+
+    /// Spawns a background task that periodically checks the cache for expired
+    /// entries and re-run checks for them.
+    ///
+    /// Removes from the cache entries that are expired and not limit order
+    /// participants and not part of the current auction. This is made to keep
+    /// addresses of long-lasting limit orders in the cache, which is the vast
+    /// majority. And addresses that open market orders which get settled pretty
+    /// quickly should get flushed out again.
+    fn spawn_maintenance_task(self: Arc<Self>) {
+        let cache_expiry = Duration::from_secs(60 * 60);
+        let maintenance_timeout = cache_expiry.div(10).max(Duration::from_secs(60));
+        let detector = Arc::clone(&self);
+
+        tokio::task::spawn(async move {
+            loop {
+                let start = Instant::now();
+                let Some(current_auction_participants) =
+                    detector.get_current_auction_participants().await
+                else {
+                    continue;
+                };
+
+                let mut expired_data: Vec<(H160, UserMetadata)> = Vec::new();
+                {
+                    let mut cache = detector.cache.write().await;
+                    let now = Instant::now();
+                    cache.retain(|address, metadata| {
+                        let expired = now
+                            .checked_duration_since(metadata.last_updated)
+                            .unwrap_or_default()
+                            >= maintenance_timeout;
+                        if expired {
+                            expired_data.push((*address, metadata.clone()));
+                        }
+                        // Remove from cache only if all the following conditions are met:
+                        // - The entry is expired
+                        // - The address is not a limit order participant
+                        // - It's not in the current auction participants
+                        !(expired
+                            && !metadata.limit_order_participant
```

Comment:
When I made the comment about long standing limit orders I didn't mean to add special handling for limit orders. I meant if the cache eviction logic removes addresses that are not in the current auction it will naturally lead to long standing limit orders remaining in the cache while quick to fill market orders get evicted after they got traded.